test/builtin_dataset_mocks.py

```diff
@@ -1344,3 +1344,96 @@ def pcam(info, root, config):
 
     return num_images
+
+
+class CityScapesMockData:
+
+    _ARCHIVE_NAMES = {
+        ("Coarse", "train"): [("gtCoarse.zip", "gtCoarse"), ("leftImg8bit_trainvaltest.zip", "leftImg8bit")],
+        ("Coarse", "train_extra"): [("gtCoarse.zip", "gtCoarse"), ("leftImg8bit_trainextra.zip", "leftImg8bit")],
+        ("Coarse", "val"): [("gtCoarse.zip", "gtCoarse"), ("leftImg8bit_trainvaltest.zip", "leftImg8bit")],
+        ("Fine", "train"): [("gtFine_trainvaltest.zip", "gtFine"), ("leftImg8bit_trainvaltest.zip", "leftImg8bit")],
+        ("Fine", "test"): [("gtFine_trainvaltest.zip", "gtFine"), ("leftImg8bit_trainvaltest.zip", "leftImg8bit")],
+        ("Fine", "val"): [("gtFine_trainvaltest.zip", "gtFine"), ("leftImg8bit_trainvaltest.zip", "leftImg8bit")],
+    }
+
+    @classmethod
+    def generate(cls, root, config):
+
+        mode = config.mode.capitalize()
+        split = config.split
+
+        if split in ["train", "train_extra"]:
+            cities = ["bochum", "bremen"]
+            num_samples = 3
+        else:
+            cities = ["bochum"]
+            num_samples = 2
+
+        polygon_target = {
+            "imgHeight": 1024,
+            "imgWidth": 2048,
+            "objects": [
+                {
+                    "label": "sky",
+                    "polygon": [
+                        [1241, 0],
+                        [1234, 156],
+                        [1478, 197],
+                        [1611, 172],
+                        [1606, 0],
+                    ],
+                },
+                {
+                    "label": "road",
+                    "polygon": [
+                        [0, 448],
+                        [1331, 274],
+                        [1473, 265],
+                        [2047, 605],
+                        [2047, 1023],
+                        [0, 1023],
+                    ],
+                },
+            ],
+        }
+
+        gt_dir = root / f"gt{mode}"
+
+        for city in cities:
+
+            def make_image(name, size=10):
+                create_image_folder(
+                    root=gt_dir / split,
+                    name=city,
+                    file_name_fn=lambda idx: name.format(idx=idx),
+                    size=size,
+                    num_examples=num_samples,
+                )
+
+            make_image(f"{city}_000000_00000" + "{idx}" + f"_gt{mode}_instanceIds.png")
+            make_image(f"{city}_000000_00000" + "{idx}" + f"_gt{mode}_labelIds.png")
+            make_image(f"{city}_000000_00000" + "{idx}" + f"_gt{mode}_color.png", size=(4, 10, 10))
+
+            for idx in range(num_samples):
+                polygon_target_name = gt_dir / split / city / f"{city}_000000_00000{idx}_gt{mode}_polygons.json"
+                with open(polygon_target_name, "w") as outfile:
+                    json.dump(polygon_target, outfile)
+
+        # Create leftImg8bit folder
+        for city in cities:
```

Comment:
We should also loop over the splits here precisely to find issues like not filtering on the split in the dataset implementation.

Reply:
Actually, I'm a bit confused about what `generate` function generates as image given that config is provided. Should it everytime generate complete dataset regardless the config or it should generate only a split ? 
From testing perspective I'd say it should generate once everything so in this case why do we have `config` as argument ?

Reply:
`generate` is just a helper construct and is not needed at all. The important part is

```py
@register_mock
def cityscapes(info, root, config):
```

This function should return all files that are needed for given `config`. In particular, every file you generate should be complete, i.e. contain everything the normal file also contains, but of course with less data. So if the regular archive has `train` / `val` / `test` folders, the generated data should have them as well, but with only a few dummy images.

> From testing perspective I'd say it should generate once everything, but in this case why do we have config as argument ?

@NicolasHug and I discussed this at length and we ended up removing the behavior for now in #5271. If you generate everything only once, you need to somehow change the return value of  the function (usually only the number of samples, but might also be more complex) for all possible configs. That made the logic more complex with (for now) little benefit. If runtime of the test ever becomes and issue, we can revisit.